Python/bytecodes.c

```diff
@@ -190,7 +190,24 @@ dummy_func(
             }
         }
 
+        op(_LOAD_BYTECODE, (--)) {
+            #ifdef Py_GIL_DISABLED
+            _Py_CODEUNIT *bytecode = _PyEval_GetExecutableCode(_PyFrame_GetCode(frame));
+            ERROR_IF(bytecode == NULL, error);
+            if (frame->bytecode != bytecode) {
```

Comment:
Since a reload of the bytecode is only necessary when the threads differ, would it make sense to store a pointer to the currently executing thread on the frame, instead of a pointer to the bytecode?

We only need the `bytecode` pointer to compute offsets, which only happens in debugging and instrumentation code.

Reply:
I'm not sure that helps simplify things. We always have to call `_PyEval_GetExecutableCode` in `RESUME`, even if the threads are the same, because it may be the first time that the thread is executing the code object (and it will need to create its copy of the bytecode).

I think we can get rid of the `bytecode` pointer if we don't care too much about the performance of `PyFrame_GetBytecode`. We could compute the `bytecode` pointer on demand by iterating through the copies of the thread-local bytecode and seeing which one contained the `instr_ptr` for the frame. This would be O(N), but maybe that's not a problem if we don't care much about perf and N is relatively small.

Reply:
I don't think we need to iterate over the array: `bytecodes = code->co_tlbc->entries[tstate->tlbc_index]`

Reply:
That doesn't work if the thread that needs to compute `bytecodes` is not the thread that is executing the frame (this happens in instrumentation). We need to retrieve the pointer to the bytecode that contains the frame's `instr_ptr` in order to compute the offset correctly. To do that we either need something on the frame (e.g. the `bytecode` pointer or the `tlbc_index` as you suggested earlier) or to loop through the copies of the bytecode for the code object and check which one contains `instr_ptr`.